pkg/tracker/vtunneltracker_test.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright © 2023 SUSE LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracker_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/docker/go-connections/nat"
+	"github.com/rancher-sandbox/rancher-desktop-agent/pkg/tracker"
+	"github.com/rancher-sandbox/rancher-desktop-agent/pkg/types"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestVTunnelTrackerAdd(t *testing.T) {
+	t.Parallel()
+
+	wslConnectAddr := []types.ConnectAddrs{{Network: "tcp", Addr: "192.168.0.1"}}
+	forwarder := testVTunnelForwarder{}
+	vtunnelTracker := tracker.NewVTunnelTracker(&forwarder, wslConnectAddr)
+
+	portMapping := nat.PortMap{
+		"80/tcp": []nat.PortBinding{
+			{
+				HostIP:   hostIP,
+				HostPort: hostPort,
+			},
+		},
+	}
+	err := vtunnelTracker.Add(containerID, portMapping)
+	assert.NoError(t, err)
+
+	portMapping2 := nat.PortMap{
+		"443/tcp": []nat.PortBinding{
+			{
+				HostIP:   hostIP2,
+				HostPort: hostPort2,
+			},
+		},
+	}
+	err = vtunnelTracker.Add(containerID2, portMapping2)
+	assert.NoError(t, err)
+
+	assert.Len(t, forwarder.receivedPortMappings, 2)
+
+	assert.ElementsMatch(t, forwarder.receivedPortMappings,
+		[]types.PortMapping{
+			{
+				Remove:       false,
+				Ports:        portMapping,
+				ConnectAddrs: wslConnectAddr,
+			}, {
+				Remove:       false,
+				Ports:        portMapping2,
+				ConnectAddrs: wslConnectAddr,
+			},
+		})
+
+	actualPortMapping1 := vtunnelTracker.Get(containerID)
+	assert.Equal(t, actualPortMapping1, portMapping)
+
+	actualPortMapping2 := vtunnelTracker.Get(containerID2)
+	assert.Equal(t, actualPortMapping2, portMapping2)
+}
+
+func TestVTunnelTrackerAddEmptyPortMap(t *testing.T) {
+	t.Parallel()
+
+	wslConnectAddr := []types.ConnectAddrs{{Network: "tcp", Addr: "192.168.0.1"}}
+	forwarder := testVTunnelForwarder{}
+	forwarder.sendErr = errSend
+	vtunnelTracker := tracker.NewVTunnelTracker(&forwarder, wslConnectAddr)
+
+	portMapping := nat.PortMap{}
+	err := vtunnelTracker.Add(containerID, portMapping)
+	assert.NoError(t, err)
+
+	assert.Len(t, forwarder.receivedPortMappings, 0)
```

Comment:
I think there's an `assert.Empty` for a better message. (But checking length zero is equivalent.)

Reply:
`assert.Empty` says `"Should be empty, but was %v"` whereas
`asser.Len` says `"\"%s\" should have %d item(s), but has %d"`
So I'm not 100% sure which is "better", I think It's just a matter of preference.